xen/arch/arm/cpufreq/wrappers.h

```diff
@@ -0,0 +1,218 @@
+#define platform_device dt_device_node
+#define of_device_id dt_device_match
+#define of_match_node dt_match_node
+#define of_property_count_elems_of_size dt_property_count_elems_of_size
+#define of_property_read_u32_index dt_property_read_u32_index
+#define of_property_for_each_string dt_property_for_each_string
+#define of_parse_phandle_with_args dt_parse_phandle_with_args
+#define of_count_phandle_with_args dt_count_phandle_with_args
+#define of_property_read_string dt_property_read_string
+#define of_parse_phandle dt_parse_phandle
+#define of_phandle_args dt_phandle_args
+#define of_get_property dt_get_property
+#define property dt_property
+
+static inline const struct of_device_id *of_match_device(
+		const struct of_device_id *matches, const struct device *dev)
+{
+	if (!matches || !dev->of_node)
+		return NULL;
+
+	return of_match_node(matches, dev->of_node);
+}
+
+/* Stuff to deal with device address ranges. */
+struct resource
+{
+	u64 start;
+	u64 size;
+};
+
+#define resource_size(res) (res)->size;
+
+static inline int of_address_to_resource(struct device_node *node, int index,
+		struct resource *res)
+{
+	return dt_device_get_address(node, index, &res->start, &res->size);
+}
+
+typedef u64 resource_size_t;
+
+#define devm_ioremap(dev, addr, size) ioremap_nocache(addr, size)
+#define devm_iounmap(dev, addr) iounmap(addr)
+
+/* Device logger functions */
+#define dev_print(dev, lvl, fmt, ...)	\
+		printk(lvl "scpi: %s: " fmt, dt_node_full_name(dev_to_dt(dev)), ## __VA_ARGS__)
+
+#define dev_info(dev, fmt, ...) dev_print(dev, XENLOG_INFO, fmt, ## __VA_ARGS__)
+#define dev_dbg(dev, fmt, ...) dev_print(dev, XENLOG_DEBUG, fmt, ## __VA_ARGS__)
+#define dev_warn(dev, fmt, ...) dev_print(dev, XENLOG_WARNING, fmt, ## __VA_ARGS__)
+#define dev_err(dev, fmt, ...) dev_print(dev, XENLOG_ERR, fmt, ## __VA_ARGS__)
+
+#define pr_debug printk
+#define _dev_info dev_info
+
+/* Helpers to get/set driver specific info. */
+static inline void platform_set_drvdata(struct platform_device *pdev, void *data)
+{
+	pdev->dev.driver_data = data;
+}
+
+static inline void *platform_get_drvdata(const struct platform_device *pdev)
+{
+	return pdev->dev.driver_data;
+}
+
+/*
+ * Xen doesn't have such a synchronization mechanism as Linux's
+ * "wait-for-completion", because of it nature. Create dummy completion
+ * infrastructure based on a busy loop to make direct ported code happy.
+ */
+struct completion {
+	atomic_t done;
+};
+
+static inline void init_completion(struct completion *x)
+{
+	atomic_set(&x->done, 0);
+}
+
+static inline void reinit_completion(struct completion *x)
+{
+	atomic_set(&x->done, 0);
+}
+
+static inline void complete(struct completion *x)
+{
+	atomic_set(&x->done, 1);
+}
+
+static inline unsigned long
+wait_for_completion_timeout(struct completion *x, unsigned long timeout)
+{
+	s_time_t deadline = NOW() + MILLISECS(timeout);
+
+	do {
+		if (atomic_cmpxchg(&x->done, 1, 0))
+			return 1;
+
+		cpu_relax();
+		udelay(1);
```

Comment:
do you really need this delay here? if we hope that cpu_relax does something then udelay may be not needed

Reply:
cpu_relax == barrier on ARM

Reply:
my point was that either you burn the CPU doing some work, e.g. checking if there is anything to do, or just burn with udelay

Reply:
Taking into the account aanisov's comment regarding a nature of wait_for_completion_timeout I am about to drop this busy loop at all. Until we find a proper solution the whole function will look like:

static inline unsigned long wait_for_completion_timeout(struct completion *x,
                                                        unsigned long timeout)
{
    if ( atomic_cmpxchg(&x->done, 1, 0) )
        return 1;

    printk("wait_for_completion isn't implemented! Return a timeout error\n");
    WARN();

    return 0;
}